src/db.c

```diff
@@ -1502,31 +1490,34 @@ int keyIsExpired(redisDb *db, robj *key) {
  * is via lookupKey*() family of functions.
  *
  * The behavior of the function depends on the replication role of the
- * instance, because slave instances do not expire keys, they wait
- * for DELs from the master for consistency matters. However even
- * slaves will try to have a coherent return value for the function,
- * so that read commands executed in the slave side will be able to
- * behave like if the key is expired even if still present (because the
- * master has yet to propagate the DEL).
+ * instance, because replicas do not expire keys. They wait for DELs
+ * from the master for consistency matters. The exception is write
+ * operations to writable replicas. However, even read-only replicas
+ * will try to have a coherent return value of the function, so that
+ * read commands executed on the replica will be able to behave like if
+ * the key is expired even if still present (because the master has yet
+ * to propagate the DEL).
  *
  * In masters as a side effect of finding a key which is expired, such
  * key will be evicted from the database. Also this may trigger the
  * propagation of a DEL/UNLINK command in AOF / replication stream.
  *
  * The return value of the function is 0 if the key is still valid,
  * otherwise the function returns 1 if the key is expired. */
-int expireIfNeeded(redisDb *db, robj *key) {
+int expireIfNeeded(redisDb *db, robj *key, int expire_on_replica) {
```

Comment:
Do we want to add an assert here?

```suggestion
int expireIfNeeded(redisDb *db, robj *key, int expire_on_replica) {
    assert(!expire_on_replica || server.masterhost == NULL || !server.repl_slave_ro);
```

Reply:
so this assertion will fire if someone calls lookupKeyWrite in a read command when that command runs on a read-only replica, right?
i'd like to catch these cases, but since we don't have much tests running on a writable replica, i don't think it'll catch them.
instead how about just asserting that `c->cmd->flags` has `CMD_WRITE`? (with a proper comment)
this will crash a lot sooner if we have such a bug.

Reply:
Considering I had to fix some test modules which marked themselves as "readonly" when they were not, we may break some other modules with this PR. Something to highlight for the release notes.

Reply:
I used `c->cmd`. What's the difference?

For the blocking commands, we can use lookupKeyRead with nostats flags to check if the key exists check its type, and delegate to the handler per type. Then, there can be another lookupKey depending on command. OK?

Reply:
> Considering I had to fix some test modules which marked themselves as "readonly" when they were not, we may break some other modules with this PR. Something to highlight for the release notes.

that may mean that we can't afford to add that assertion. we need to be ABI compatible with old modules. maybe exclude module commands from that assertion?

Reply:
> OK, thx. So, I can confirm that during exec, server.current_client->cmd is the queued command, but when handleClientsBlockedOnKeys is called afterwards, server.current_client->cmd is exec.

ok, i didn't realize that we're discussing blocked clients here too... but then i don't understand how this is related to exec (blocking commands aren't working in a transaction.

> Although the command function (e.g. blockingGenericZpopCommand) is passed a client, lookupKey and expireIfNeeded aren't. They only have access to server.current_client, which is not the same client in this case. A test case which crashes is "MULTI/EXEC is isolated from the point of view of $pop" in unit/zset.tcl.

correct me if i'm wrong, but it's not just the fact the current_client is not set, and the client is not passed to lookupKey. the problem is that the lookupKey is done before looping on the list of clients that are blocked on that key (so it's one lookup possibly for serving many clients and many different commands). in that case the only way out IMHO is to tell lookupKey not to record stats, and instead have the individual blocked command handlers update the stats later.

> One idea is to set server.current_client to the client which is being served (running ZPOPMIN, BLMOVE, etc.) so that the assert in expireIfNeeded works, and change it back to the old current_client (the one which triggered the unblocking using exec or eval) afterwards. It feels like a hack though.

please note that in unstable this is already handled (#9422), i think you should update your branch.

> For debug loadaof, similarily, maybe it's possible to temporarily set server.current_client to the fake client which executes commands from the AOF file.

seems right, please go ahead.

> Setting current_client seem like a hack, right? Passing the client around to lookupKey and expireIfNeeded might be a better idea but it's a lot of work to do it everywhere. Another option is to just scrap the assert...

yes, the use of globals in redis may be a problem some day, but let's not deal with it now.
i think this assert may be very useful, if we write any code that relies on these flags, this assert will give us an indication that we're not introducing bugs.

Reply:
> but then i don't understand how this is related to exec (blocking commands aren't working in a transaction.

The blocked client (e.g. BLPOP) is not in a transaction. It's another client which unblocks the blocked client by e.g. RPUSH in a transaction (exec) or Lua (eval).

> in unstable this [setting server.current_client] is already handled

Nice! I didn't see that. I will update and then do the same for debug loadaof.

(I was just starting to implement a lookupKeyForClient with an extra client parameter, only for use when the client is not the current client. The old lookup functions would remain and use server.current_client as a default. But I didn't get that far.)

Reply:
I think we have to exempt modules from that assertion.
the new redis release needs to be compatible with old modules, we need to consider the implications of modules using the wrong lookup.

The main difference in REDISMODULE_WRITE, is that RM_OpenKey returns a non-null RedisModuleKey even if the key is not found in redis, in which case `kp->value` will be NULL.
And modules that do that, expect that and could still use RM_ModuleTypeSetValue and RM_ModuleTypeReplaceValue, and unlike redis (e.g. incrDecrCommand) don't attempt to access the robj directly without checks.

So let's try to consider the implications of our change (assuming the assertion is skipped).
We moved the expiration attempt from lookupKeyRead to lookupKey, so it now affects REDISMODULE_WRITE too.
And the difference is only when used on a replica (readonly or writable, since modules could use that on RO commands), when a command is executed from the non-master client.
And the implication would be that both before and after our fix, RM_OpenKey succeeds, however after our fix, any attempt to access the value RM_HashGet will fail (only if the key already expired), and in the past it would have succeeded.
Either way, with or without our fix, RM_ModuleTypeSetValue and RM_ModuleTypeReplaceValue will succeed and do the same thing.

anything i'm missing?

regarding BLPOP, can you please describe the flow in the code itself?
AFAICT `handleClientsBlockedOnKeys` is only called by `processCommand` so would only be called when EVAL returns.